src/ComponentProvider.ts

```diff
@@ -1,94 +1,60 @@
-      }
-
-      if (channel.hasSubscribe()) {
-        const message = channel.subscribe().message();
-        const currentPath = `${path}.subscribe.message`;
-        this.scanMessage(currentPath, message);
-        this.messages.set(currentPath, message);
-      }
-
-      if (channel.hasParameters()) {
-        const currentPath = `${path}.parameters`;
-        for (const [name, channelParameter] of Object.entries(channel.parameters())) {
-          this.parameters.set(`${currentPath}.${name}`, channelParameter);
-        }
-      }
-    }
-  }
-
-  private scanSchema(path: string, schema: Schema): void {
-    if (!schema) {return;}
-    this.schemas.set(path, schema);
-    const schemaProperties = schema.properties();
-    for (const [propertyName, propertySchema] of Object.entries(schemaProperties)) {
-      this.schemas.set(`${path}.properties.${propertyName}`, propertySchema);
-    }
-  }
-
-  private scanMessageTraits(path: string, traits: MessageTrait[]): void {
-    for (const [index, trait] of Object.entries(traits)) {
-      if (trait.headers()) {
-        this.scanSchema(`${path}[${index}].headers`, trait.headers());
-      }
-    }
+    this.messages = this.parseComponents(this.messagePaths);
+    this.schemas = this.parseComponents(this.schemaPaths);
+    this.parameters = this.parseComponents(this.parameterPaths);
   }
 
-  private scanMessage(path: string, message: Message): void {
-    this.scanSchema(`${path}.payload`, message.payload());
-    this.scanSchema(`${path}.headers`, message.headers());
-    this.scanMessageTraits(`${path}.traits`, message.traits());
+  private toLodashPath(JSONPath: string) {
+    return JSONPath.replace(/'\]\['/g, '.')
+      .slice(3, -2)
+      .replace(/'\]/g, '')
+      .replace(/\['/g, '.');
   }
-
-  private scanComponents(): void {
-    const components = this.document.components();
-    if (!components) {
-      return;
-    }
-
-    if (components.hasMessages()) {
-      for (const [messageName, message] of Object.entries(components.messages())) {
-        this.messages.set(`components.messages.${messageName}`, message);
-      }
-    }
-
-    if (components.hasSchemas()) {
-      for (const [schemaName, schema] of Object.entries(components.schemas())) {
-        this.schemas.set(`components.schemas.${schemaName}`, schema);
-      }
-    }
-
-    if (components.hasParameters()) {
-      for (const [parameterName, parameter] of Object.entries(components.parameters())) {
-        this.parameters.set(`components.parameters.${parameterName}`, parameter);
-      }
-    }
+  private parseComponents(paths: string[]): any {
+    return _.chain(paths)
+      .map((messagePath) => {
+        return JSONPath({
+          path: messagePath,
+          resultType: 'all',
+          json: this.document.json(),
+        });
+      })
+      .flatten()
+      .reduce((red, message) => {
+        return red.set(this.toLodashPath(message.path), message.value);
+      }, new Map())
+      .value();
```

Comment:
`value()` function on the `Map`? Maybe I don't understand something 😅 

```suggestion
      .values();
```

Reply:
`.value()` is for the lodash `chain` method. https://lodash.com/docs/4.17.15#chain